service/src/main/kotlin/fi/espoo/evaka/pairing/PairingQueries.kt

```diff
@@ -0,0 +1,125 @@
+package fi.espoo.evaka.pairing
+
+import fi.espoo.evaka.shared.db.Database
+import fi.espoo.evaka.shared.domain.NotFound
+import fi.espoo.evaka.shared.utils.zoneId
+import org.jdbi.v3.core.kotlin.mapTo
+import java.security.SecureRandom
+import java.time.ZonedDateTime
+import java.util.UUID
+
+const val maxAttempts = 10
+
+fun initPairing(tx: Database.Transaction, unitId: UUID): Pairing {
+    // language=sql
+    val sql =
+        """
+            INSERT INTO pairing (unit_id, expires, challenge_key) 
+            VALUES (:unitId, :expires, :challenge)
+            RETURNING *
+        """.trimIndent()
+
+    return tx.createQuery(sql)
+        .bind("unitId", unitId)
+        .bind("expires", ZonedDateTime.now(zoneId).plusMinutes(15).toInstant())
+        .bind("challenge", generatePairingKey())
+        .mapTo<Pairing>()
+        .first()
+}
+
+fun challengePairing(tx: Database.Transaction, challengeKey: String): Pairing {
+    // language=sql
+    val sql =
+        """
+            UPDATE pairing SET response_key = :response, status = 'WAITING_RESPONSE'
+            WHERE challenge_key = :challenge AND status = 'WAITING_CHALLENGE' AND expires > :now AND attempts <= :maxAttempts
+            RETURNING *
+        """.trimIndent()
+
+    return tx.createQuery(sql)
+        .bind("challenge", challengeKey)
+        .bind("response", generatePairingKey())
+        .bind("now", ZonedDateTime.now(zoneId).toInstant())
+        .bind("maxAttempts", maxAttempts)
+        .mapTo<Pairing>()
+        .list()
```

Comment:
Call to .list() is unnecessary if we are only interested in first element (similar below)